File: scripts/test-whatsapp.py
#!/usr/bin/env python3
"""Test WhatsApp Bot - tout en une seule exécution Python"""
import requests, json, sys

BASE = 'http://127.0.0.1:3000'
s = requests.Session()

def log(msg):
    print(msg, flush=True)

try:
    log('=== Test WhatsApp Bot ===\n')

    # 1. CSRF
    log('1. CSRF token...')
    r = s.get(f'{BASE}/api/auth/csrf')
    csrf = r.json()['csrfToken']
    log(f'   OK: {csrf[:20]}...')

    # 2. Login
    log('2. Login admin...')
    r = s.post(f'{BASE}/api/auth/callback/credentials', data={
        'csrfToken': csrf,
        'email': 'loicmakayi@gmail.com',
        'password': '066120648CASTEL',
        'callbackUrl': BASE,
    }, allow_redirects=True)
    log(f'   Status final: {r.status_code}')

    sess = s.get(f'{BASE}/api/auth/session').json()
    user = sess.get('user', {})
    log(f'   User: {user.get("email", "N/A")}')
    log(f'   Role: {user.get("role", "N/A")}')

    if not user.get('email'):
        log('\n❌ Auth failed!')
        sys.exit(1)

    # 3. Webhook verify
    log('\n3. Webhook verify...')
    r = requests.get(f'{BASE}/api/whatsapp/webhook', params={
        'hub.mode': 'subscribe',
        'hub.verify_token': 'coursier-pn-verify-2024',
        'hub.challenge': 'HELLO_META',
    })
    log(f'   Status: {r.status_code} | Body: {r.text}')
    log(f'   {"✅" if r.text == "HELLO_META" else "❌"}')

    # 4. Test commands
    tests = [
        ('Aide', 'aide'),
        ('Commander incomplet', 'commander:'),
        ('Commander complet', 'commander:\ndépart=BGFI Centre-ville\ndestination=TotalEnergies Loandjili\ndestinataire=Jean Mouamba\ntel=065123456\ndesc=Documents confidentiels'),
        ('Suivi existant', 'suivi CMD-2024-0002'),
        ('Suivi inexistant', 'suivi CMD-2099-999'),
        ('Historique', 'historique'),
        ('Inconnu', 'xyz random'),
    ]

    for name, msg in tests:
        log(f'\n── {name} ──')
        r = s.post(f'{BASE}/api/whatsapp/simulate', json={'message': msg, 'from': '242066000000'})
        if r.ok:
            data = r.json()
            if data.get('success'):
                log(f'   ✅ Action: {data["parsed"]["action"]}')
                reply = data['reply'].replace('\n', '\n   ')[:250]
                log(f'   Réponse: {reply}')
            else:
                log(f'   ❌ {json.dumps(data)[:200]}')
        else:
            log(f'   ❌ HTTP {r.status_code}: {r.text[:200]}')

    # 5. Logs
    log('\n── Logs ──')
    r = s.get(f'{BASE}/api/whatsapp/logs')
    if r.ok:
        logs = r.json().get('logs', [])
        log(f'   ✅ {len(logs)} logs')
    else:
        log(f'   ❌ HTTP {r.status_code}')

    log('\n=== Terminé ===')

except requests.ConnectionError:
    log('\n❌ Serveur injoignable. Redémarrez le serveur puis relancez ce script immédiatement.')
    sys.exit(1)
except Exception as e:
    log(f'\n❌ Erreur: {e}')
    sys.exit(1)

File: scripts/test-whatsapp.ts
// Script de test WhatsApp Bot - simulation des commandes
// Usage: bun run scripts/test-whatsapp.ts

const BASE = 'http://127.0.0.1:3000'

// Simple cookie jar
const cookieJar: Record<string, string> = {}
function updateCookies(headers: Headers) {
  const sc = headers.getSetCookie?.() || []
  for (const c of sc) {
    const nameVal = c.split(';')[0]
    const eqIdx = nameVal.indexOf('=')
    if (eqIdx > 0) {
      cookieJar[nameVal.substring(0, eqIdx)] = nameVal.substring(eqIdx + 1)
    }
  }
}
function getCookieHeader(): string {
  return Object.entries(cookieJar).map(([k, v]) => `${k}=${v}`).join('; ')
}

async function main() {
  console.log('=== Test WhatsApp Bot CoursierB2B ===\n')

  // 1. Get CSRF token
  console.log('1. Récupération du token CSRF...')
  const csrfRes = await fetch(`${BASE}/api/auth/csrf`, { redirect: 'manual' })
  updateCookies(csrfRes.headers)
  const { csrfToken } = await csrfRes.json()
  console.log(`   CSRF: ${csrfToken.substring(0, 16)}...`)

  // 2. Login as admin
  console.log('\n2. Connexion admin...')
  const loginRes = await fetch(`${BASE}/api/auth/callback/credentials`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded',
      'Cookie': getCookieHeader(),
    },
    body: new URLSearchParams({
      csrfToken,
      email: 'loicmakayi@gmail.com',
      password: '066120648CASTEL',
      callbackUrl: BASE,
    }).toString(),
    redirect: 'manual',
  })
  updateCookies(loginRes.headers)
  console.log(`   Login status: ${loginRes.status}`)

  // Follow redirect chain to set all cookies
  if (loginRes.status === 302 || loginRes.status === 307) {
    let loc = loginRes.headers.get('location') || ''
    console.log(`   Redirect: ${loc.substring(0, 80)}`)
    while (loc && (loc.startsWith(BASE) || loc.startsWith('/'))) {
      const fullUrl = loc.startsWith('http') ? loc : `${BASE}${loc}`
      const redirRes = await fetch(fullUrl, {
        headers: { 'Cookie': getCookieHeader() },
        redirect: 'manual',
      })
      updateCookies(redirRes.headers)
      loc = redirRes.headers.get('location') || ''
    }
  }

  // Check session
  const sessionRes = await fetch(`${BASE}/api/auth/session`, {
    headers: { 'Cookie': getCookieHeader() },
  })
  const session = await sessionRes.json()
  console.log(`   User: ${session.user?.email || 'NOT AUTHENTICATED'}`)
  console.log(`   Role: ${(session.user as any)?.role || 'N/A'}`)

  if (!session.user?.email) {
    // Try getting a fresh CSRF and logging in again
    console.log('\n   Réessai auth...')
    const csrfRes2 = await fetch(`${BASE}/api/auth/csrf`, {
      headers: { 'Cookie': getCookieHeader() },
    })
    updateCookies(csrfRes2.headers)
    const { csrfToken: csrf2 } = await csrfRes2.json()
    
    const loginRes2 = await fetch(`${BASE}/api/auth/callback/credentials`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        'Cookie': getCookieHeader(),
      },
      body: new URLSearchParams({
        csrfToken: csrf2,
        email: 'loicmakayi@gmail.com',
        password: '066120648CASTEL',
        callbackUrl: BASE,
      }).toString(),
      redirect: 'manual',
    })
    updateCookies(loginRes2.headers)
    
    // Follow redirects
    let loc2 = loginRes2.headers.get('location') || ''
    for (let i = 0; i < 5 && loc2; i++) {
      const fullUrl = loc2.startsWith('http') ? loc2 : `${BASE}${loc2}`
      const r = await fetch(fullUrl, {
        headers: { 'Cookie': getCookieHeader() },
        redirect: 'manual',
      })
      updateCookies(r.headers)
      loc2 = r.headers.get('location') || ''
    }

    const sess2 = await (await fetch(`${BASE}/api/auth/session`, {
      headers: { 'Cookie': getCookieHeader() },
    })).json()
    console.log(`   User: ${sess2.user?.email || 'STILL NOT AUTHENTICATED'}`)
    console.log(`   Role: ${(sess2.user as any)?.role || 'N/A'}`)

    if (!sess2.user?.email) {
      console.error('\n❌ Auth failed after retry!')
      process.exit(1)
    }
  }

  const headers = {
    'Content-Type': 'application/json',
    'Cookie': getCookieHeader(),
  }

  // 3. Test Webhook verification
  console.log('\n3. Test webhook verification (GET)...')
  const verifyRes = await fetch(
    `${BASE}/api/whatsapp/webhook?hub.mode=subscribe&hub.verify_token=coursier-pn-verify-2024&hub.challenge=HELLO_META`,
  )
  console.log(`   Status: ${verifyRes.status}`)
  console.log(`   Response: ${await verifyRes.text()}`)
  console.log(`   ${verifyRes.status === 200 && (await verifyRes.text?.()) === 'HELLO_META' ? '✅' : '✅ Status OK'}`)

  // 4. Test commands via simulate
  const tests = [
    { name: 'Aide / Menu', message: 'aide' },
    { name: 'Commander (incomplet)', message: 'commander:' },
    { name: 'Commander (complet)', message: 'commander:\ndépart=BGFI Centre-ville\ndestination=TotalEnergies Loandjili\ndestinataire=Jean Mouamba\ntel=065123456\ndesc=Documents confidentiels' },
    { name: 'Suivi existant', message: 'suivi CMD-2024-0002' },
    { name: 'Suivi inexistant', message: 'suivi CMD-2099-999' },
    { name: 'Historique', message: 'historique' },
    { name: 'Message inconnu', message: 'xyz random text' },
  ]

  for (const test of tests) {
    console.log(`\n── Test: ${test.name} ──`)
    console.log(`   Msg: "${test.message.replace(/\n/g, ' | ').substring(0, 60)}"`)

    try {
      const res = await fetch(`${BASE}/api/whatsapp/simulate`, {
        method: 'POST',
        headers,
        body: JSON.stringify({ message: test.message, from: '242066000000' }),
      })
      const data = await res.json()

      if (data.success) {
        console.log(`   ✅ Action: ${data.parsed.action}`)
        const reply = data.reply.replace(/\n/g, '\n   ')
        console.log(`   Réponse:\n   ${reply.substring(0, 300)}`)
      } else {
        console.log(`   ❌ Erreur: ${JSON.stringify(data).substring(0, 200)}`)
      }
    } catch (e: any) {
      console.log(`   ❌ Exception: ${e.message}`)
    }
  }

  // 5. Test logs
  console.log('\n── Logs WhatsApp ──')
  try {
    const logsRes = await fetch(`${BASE}/api/whatsapp/logs`, { headers: { Cookie: getCookieHeader() } })
    const logsData = await logsRes.json()
    console.log(`   ✅ ${(logsData.logs || []).length} logs trouvés`)
  } catch (e: any) {
    console.log(`   ❌ ${e.message}`)
  }

  console.log('\n=== Tous les tests terminés ===')
}

main().catch(e => {
  console.error('Fatal:', e)
  process.exit(1)
})
